src/ripple/shamap/SHAMapTreeNode.h

```diff
@@ -164,67 +215,70 @@ SHAMapTreeNode::isLeaf () const
 
 inline
 bool
-SHAMapTreeNode::isInner () const
+SHAMapAbstractNode::isInner () const
 {
     return mType == tnINNER;
 }
 
 inline
 bool
-SHAMapTreeNode::isInBounds (SHAMapNodeID const &id) const
+SHAMapAbstractNode::isValid () const
 {
-    // Nodes at depth 64 must be leaves
-    return (!isInner() || (id.getDepth() < 64));
+    return mType != tnERROR;
 }
 
+// SHAMapInnerNode
+
 inline
-bool
-SHAMapTreeNode::isValid () const
+SHAMapInnerNode::SHAMapInnerNode(std::uint32_t seq)
+    : SHAMapAbstractNode(tnINNER, seq)
 {
-    return mType != tnERROR;
 }
 
 inline
 bool
-SHAMapTreeNode::isTransaction () const
+SHAMapInnerNode::isEmptyBranch (int m) const
 {
-    return (mType == tnTRANSACTION_NM) || (mType == tnTRANSACTION_MD);
+    return (mIsBranch & (1 << m)) == 0;
 }
 
 inline
-bool
-SHAMapTreeNode::hasMetaData () const
+uint256 const&
+SHAMapInnerNode::getChildHash (int m) const
 {
-    return mType == tnTRANSACTION_MD;
+    assert ((m >= 0) && (m < 16) && (getType() == tnINNER));
+    return mHashes[m];
 }
 
 inline
 bool
-SHAMapTreeNode::isAccountState () const
+SHAMapInnerNode::isFullBelow (std::uint32_t generation) const
 {
-    return mType == tnACCOUNT_STATE;
+    return mFullBelowGen == generation;
 }
 
 inline
-bool
-SHAMapTreeNode::isInnerNode () const
+void
+SHAMapInnerNode::setFullBelowGen (std::uint32_t gen)
 {
-    return !mItem;
+    mFullBelowGen = gen;
 }
 
 inline
 bool
-SHAMapTreeNode::isEmptyBranch (int m) const
+SHAMapAbstractNode::isInBounds (SHAMapNodeID const &id) const
```

Comment:
It's not necessarily intrinsically wrong, but this `SHAMapAbstractNode` function implementation is mixed in with `SHAMapInnerNode` implementations, which may cause some confusion.